salt/states/ansiblegate.py

```diff
@@ -99,3 +103,88 @@ def __virtual__():
     '''
     setattr(sys.modules[__name__], 'call', lambda **kwargs: AnsibleState()(**kwargs))   # pylint: disable=W0108
     return ansible is not None
+
+
+def _client():
+    '''
+    Get a fileclient
+    '''
+    return salt.fileclient.get_file_client(__opts__)
+
+
+def _changes(plays):
+    '''
+    Find changes in ansible return data
+    '''
+    changes = {}
+    for play in plays['plays']:
+        task_changes = {}
+        for task in play['tasks']:
+            host_changes = {}
+            for host, data in six.iteritems(task['hosts']):
+                if data['changed'] is True:
+                    host_changes[host] = data.get('diff', data.get('changes', {}))
+            if host_changes:
+                task_changes[task['task']['name']] = host_changes
+        if task_changes:
+            changes[play['play']['name']] = task_changes
+    return changes
+
+
+def playbooks(name, rundir=None, gitrepo=None, git_kwargs=None, ansible_kwargs=None):
+    '''
+    Run Ansible Playbooks
+
+    :param name: path to playbook. This can be relative to rundir or the git repo
+    :param rundir: location to run ansible-playbook from.
+    :param gitrepo: gitrepo to clone for ansible playbooks.  This is cloned
+                    using the `git.latest` state, and is cloned to the `rundir`
+                    if specified, otherwise it is clone to the `cache_dir`
+    :param git_kwargs: extra kwargs to pass to `git.latest` state module besides
+                       the `name` and `target`
+    :param ansible_kwargs: extra kwargs to pass to `ansible.playbooks` execution
+                           module besides the `name` and `target`
+
+    :return: Ansible playbook output.
+
+    .. code-block:: yaml
+
+        run nginx install:
+          ansible.playbooks:
+            - name: install.yml
+            - gitrepo: git://github.com/gituser/playbook.git
+            - git_kwargs:
+                rev: master
+    '''
+    ret = {
+        'result': False,
+        'changes': {},
+        'comment': 'Running playbook {0}'.format(name),
+        'name': name,
+    }
+    if gitrepo is not None:
+        if rundir is None:
+            rundir = _client()._extrn_path(gitrepo, 'base')
+        if git_kwargs is None:
+            git_kwargs = {}
+        __states__['git.latest'](
+            name=gitrepo,
+            target=rundir,
+            **git_kwargs,
+        )
+    if ansible_kwargs is None:
+        ansible_kwargs = {}
```

Comment:
Maybe this is better way, unless we do not want to validate all that and crash with the error "You've passed an empty string instead of a parameters" (or something like that):

```python
if not ansible_kwargs:
    ansible_kwargs = {}
```

Reply:
I am just going to check if it a dict and if it isn't then i will set it to {}.